File: C++/Queue/Queue_06.cpp
/*
===============================================================================

Program Name    : Queue Complete Program
Program Number  : 006
File Name       : Queue_06.cpp

Description     : This program implements Enqueue, Dequeue, Display
                  and Count operations of Queue using a singly
                  linear linked list.
                  
Date            : August 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve

===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

class Queue
{
private:
    struct node *first;
    int iCount;

public:
    Queue();
    void Enqueue(int iNo);
    int Dequeue();
    void Display();
    int Count();
};

///////////////////////////////////////////////////////////////////////////////
//
// Constructor
//
///////////////////////////////////////////////////////////////////////////////

Queue :: Queue()
{
    this->first = NULL;
    this->iCount = 0;
}

///////////////////////////////////////////////////////////////////////////////
//
// Function Name : Enqueue
// Description   : Insert element at Last position
//
///////////////////////////////////////////////////////////////////////////////

void Queue :: Enqueue(int iNo)
{
    struct node *newn = NULL;
    struct node *temp = NULL;

    newn = new struct node();

    newn->data = iNo;
    newn->next = NULL;

    if(first == NULL)
    {
        first = newn;
    }
    else
    {
        temp = first;

        while(temp->next != NULL)
        {
            temp = temp->next;
        }

        temp->next = newn;
    }

    iCount++;
}

///////////////////////////////////////////////////////////////////////////////
//
// Function Name : Dequeue
// Description   : Delete element from First position
//
///////////////////////////////////////////////////////////////////////////////

int Queue :: Dequeue()
{
    int iValue = 0;
    struct node *temp = NULL;

    if(first == NULL)
    {
        cout<<"Queue is Empty!!!\n";
        return -1;
    }

    iValue = first->data;

    temp = first;

    first = first->next;

    delete temp;

    iCount--;

    return iValue;
}

///////////////////////////////////////////////////////////////////////////////
//
// Function Name : Display
// Description   : Display Queue elements
//
///////////////////////////////////////////////////////////////////////////////

void Queue :: Display()
{
    struct node *temp = NULL;

    temp = first;

    cout<<"\nQueue Elements : \n";

    while(temp != NULL)
    {
        cout<<"| "<<temp->data<<" |"<<endl;
        temp = temp->next;
    }
}

///////////////////////////////////////////////////////////////////////////////
//
// Function Name : Count
// Description   : Return number of elements
//
///////////////////////////////////////////////////////////////////////////////

int Queue :: Count()
{
    return iCount;
}

///////////////////////////////////////////////////////////////////////////////
//
// Entry Point Function
//
///////////////////////////////////////////////////////////////////////////////

int main()
{
    Queue qobj;
    int iRet = 0;

    qobj.Enqueue(11);
    qobj.Enqueue(21);
    qobj.Enqueue(51);
    qobj.Enqueue(101);

    qobj.Display();

    iRet = qobj.Count();

    cout<<"\nNumber of Elements in Queue are : "<<iRet<<endl;

    iRet = qobj.Dequeue();

    cout<<"\nRemoved Element : "<<iRet<<endl;

    qobj.Display();

    iRet = qobj.Count();

    cout<<"\nNumber of Elements in Queue are : "<<iRet<<endl;

    iRet = qobj.Dequeue();

    cout<<"\nRemoved Element : "<<iRet<<endl;

    qobj.Display();

    iRet = qobj.Count();

    cout<<"\nNumber of Elements in Queue are : "<<iRet<<endl;

    return 0;
}

/*
===============================================================================
Expected Output
===============================================================================

Queue Elements :
| 11 |
| 21 |
| 51 |
| 101 |

Number of Elements in Queue are : 4

Removed Element : 11

Queue Elements :
| 21 |
| 51 |
| 101 |

Number of Elements in Queue are : 3

Removed Element : 21

Queue Elements :
| 51 |
| 101 |

Number of Elements in Queue are : 2

===============================================================================
Algorithm
===============================================================================

1. Start.
2. Create a Queue object.
3. Initialize first to NULL and iCount to 0.
4. Insert 11, 21, 51 and 101 using Enqueue().
5. Display all Queue elements using Display().
6. Call Count() to get the total number of elements.
7. Call Dequeue() to remove the first element.
8. Store the data of the first node.
9. Move first to the next node.
10. Delete the previous first node.
11. Decrease iCount by 1.
12. Display the removed element.
13. Display the updated Queue.
14. Count and display the remaining elements.
15. Perform Dequeue() again.
16. Display the updated Queue and count.
17. Stop.

===============================================================================
Program Flow
===============================================================================

Start
  |
  v
Create Queue Object
  |
  v
Initialize first = NULL
iCount = 0
  |
  v
Enqueue 11, 21, 51, 101
  |
  v
Display Queue
  |
  v
Count Elements
  |
  v
Dequeue First Element
  |
  v
Delete First Node
  |
  v
Decrease iCount
  |
  v
Display Updated Queue
  |
  v
Dequeue Again
  |
  v
Display Updated Queue
  |
  v
Count Remaining Elements
  |
  v
Stop

===============================================================================
Time Complexity
===============================================================================

| Operation | Time Complexity |
|-----------|-----------------|
| Enqueue() | O(n) |
| Dequeue() | O(1) |
| Display() | O(n) |
| Count() | O(1) |

===============================================================================
Space Complexity
===============================================================================

O(n)

Where n is the number of elements present in the Queue.

===============================================================================
Conclusion
===============================================================================

This program successfully implements the Queue data structure using a
singly linear linked list. It demonstrates insertion using Enqueue(),
deletion using Dequeue(), displaying elements using Display(), and
maintaining the number of elements using Count(). The Queue follows the
FIFO (First In First Out) principle.

===============================================================================
*/

File: Java/Queue/Queue_06.java
/*
===============================================================================
Program Name    : Queue - Complete Application
Program Number  : 06
File Name       : Queue_06.java

Description     : This program demonstrates multiple Dequeue operations of a
                  Queue using a singly linear linked list. Elements are
                  removed from the first position of the Queue according
                  to the FIFO (First In First Out) principle.

Date            : August 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class
2. Object-Oriented Programming (OOP)
3. Self-Referential Class
4. Constructor
5. Singly Linear Linked List
6. Queue
7. Enqueue Operation
8. Dequeue Operation
9. Display Operation
10. Count Operation
11. FIFO Principle

Key Learning Points:
- Understand multiple Dequeue operations.
- Learn how elements are removed from the first position.
- Understand the FIFO principle of Queue.
- Learn how the first pointer changes after deletion.
- Understand how iCount is updated after every Dequeue operation.
- Display the Queue after each deletion.
- Track the number of remaining Queue elements.

===============================================================================
*/

class node
{
    public int data;
    public node next;
}

class Queue
{
    private node first;
    private int iCount;

    public Queue()
    {
        first = null;
        iCount = 0;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Enqueue
    // Description   : Insert element at Last
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Enqueue(int iNo)
    {
        node newn = null;
        node temp = null;

        newn = new node();

        newn.data = iNo;
        newn.next = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            temp = first;

            while(temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
        }

        iCount++;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Dequeue
    // Description   : Delete element from First
    //
    ///////////////////////////////////////////////////////////////////////////

    public int Dequeue()
    {
        int iValue = 0;
        node temp = null;

        if(first == null)
        {
            System.out.println("Queue is Empty");
            return -1;
        }

        iValue = first.data;

        temp = first;
        first = first.next;
        temp = null;

        iCount--;

        return iValue;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Display
    // Description   : Display Queue
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Display()
    {
        node temp = first;

        System.out.println("\nElements of Queue are : ");

        while(temp != null)
        {
            System.out.print("|"+temp.data+"| -> ");
            temp = temp.next;
        }

        System.out.println("NULL");
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Count
    // Description   : Count number of elements
    //
    ///////////////////////////////////////////////////////////////////////////

    public int Count()
    {
        return iCount;
    }
}

class Queue_06
{
    public static void main(String A[])
    {
        Queue qobj = new Queue();

        int iRet = 0;

        qobj.Enqueue(11);
        qobj.Enqueue(21);
        qobj.Enqueue(51);
        qobj.Enqueue(101);

        qobj.Display();

        iRet = qobj.Count();

        System.out.println("\nNumber of Elements : "+iRet);

        iRet = qobj.Dequeue();

        System.out.println("\nRemoved Element : "+iRet);

        qobj.Display();

        iRet = qobj.Count();

        System.out.println("\nNumber of Elements : "+iRet);

        iRet = qobj.Dequeue();

        System.out.println("\nRemoved Element : "+iRet);

        qobj.Display();

        iRet = qobj.Count();

        System.out.println("\nNumber of Elements : "+iRet);
    }
}

/*
===============================================================================
Expected Output:
===============================================================================

Elements of Queue are :
|11| -> |21| -> |51| -> |101| -> NULL

Number of Elements : 4

Removed Element : 11

Elements of Queue are :
|21| -> |51| -> |101| -> NULL

Number of Elements : 3

Removed Element : 21

Elements of Queue are :
|51| -> |101| -> NULL

Number of Elements : 2

===============================================================================
Memory Representation:
===============================================================================

Initial Queue:

first
  |
  v
+------+-------+    +------+-------+    +------+-------+    +------+-------+
|  11  |   o-------->|  21  |   o-------->|  51  |   o-------->| 101  | NULL|
+------+-------+    +------+-------+    +------+-------+    +------+-------+


After First Dequeue:

first
  |
  v
+------+-------+    +------+-------+    +------+-------+
|  21  |   o-------->|  51  |   o-------->| 101  | NULL|
+------+-------+    +------+-------+    +------+-------+


After Second Dequeue:

first
  |
  v
+------+-------+    +------+-------+
|  51  |   o-------->| 101  | NULL|
+------+-------+    +------+-------+

===============================================================================
Algorithm:
===============================================================================

1. Start the program.
2. Create a Queue object.
3. Initialize first to NULL and iCount to 0.
4. Insert 11, 21, 51 and 101 using Enqueue().
5. Display the Queue elements.
6. Count and display the number of Queue elements.
7. Call Dequeue() for the first time.
8. Store the data of the first node.
9. Move first to the next node.
10. Decrease iCount.
11. Display the removed element.
12. Display the updated Queue.
13. Display the updated count.
14. Call Dequeue() for the second time.
15. Remove the current first node.
16. Move first to the next node.
17. Decrease iCount.
18. Display the removed element.
19. Display the updated Queue.
20. Display the updated count.
21. Stop the program.

===============================================================================
Program Flow:
===============================================================================

        Start
          |
          v
    Create Queue Object
          |
          v
    Enqueue 4 Elements
          |
          v
     Display Queue
          |
          v
       Count()
          |
          v
      Dequeue 1
          |
          v
    Update first Pointer
          |
          v
     Display Queue
          |
          v
       Count()
          |
          v
      Dequeue 2
          |
          v
    Update first Pointer
          |
          v
     Display Queue
          |
          v
       Count()
          |
          v
         Stop

===============================================================================
Time Complexity:
===============================================================================

Enqueue() : O(n)
Dequeue() : O(1)
Display() : O(n)
Count()   : O(1)

===============================================================================
Space Complexity:
===============================================================================

O(n) total space for n Queue nodes.

===============================================================================
Conclusion:
===============================================================================

Program 06 demonstrates multiple Dequeue operations in a Queue using a
singly linear linked list in Java. The elements 11 and 21 are removed
sequentially from the front of the Queue, demonstrating the FIFO
(First In First Out) principle. The first pointer and iCount are updated
after every successful Dequeue operation.

===============================================================================
*/
